### DISCUSSION TITLE
Is there an helper atom function like derivedWithEqualityAtom?

### DISCUSSION THREAD

**Discussion by wootra** [pmndrs/jotai#1934] [Q&A] [answered]
Hello, team.
while we have selectAtom, we can reduce rendering by using equalityFn in selectAtom, I wonder if there is such thing that can reduce rendering from derived atom?
If I use selectAtom and derived atom at the same time, the code would be like this:
```
const someAtom = atom({ a: 1, b: 1});
const derivedWithEqualityAtom = selectAtom( atom( get=>{
  const obj = get(someAtom);
  return {
    a: obj.a,
    b: obj.b,
    sum: obj.a + obj.b,
  }
}), state=>state, compareObj);
```
it would be still simple, but I think it can be simpler if we have something like this:

```
const someAtom =  derivedAtom( get=> { ... }, compareObj);
```

and derivedAtom will look like this:
```
export const derivedAtom = <T>(read: (get: Getter) => T, equalityFn = Object.is) => {
	return selectAtom(atom(read), (state) => state, equalityFn);
};

```

on top of that, my coworker brought an interesting perspective about selectAtom and derived atom.
since selectAtom's name is 'select' atom, it would be used to take a slice of an atom instead of making a new object ( like above example ). in that case, even though it is derived from 1 atom, we should use derived atom. I like the concept itself, but want to know what do you think about this perspective. 

Thanks!

**Comment by dai-shi** [comment]
Using custom equality function is not recommended in general and we discourage the usage.
However, it's sometimes valid, and thus `selectAtom` is provided as an escape hatch.

Feel free to create such a helper function on your end if your team agrees on such a usage.

**Reply by wootra** [reply]
thank you for the answer @dai-shi ! the compareObj we are using is shallow compare which just check the 1st level of the object field that is given to selectAtom since in above case, derive atom will always create a new object reference that makes Object.is always returns false. 

I have another question for your opinion. 
while selectAtom still can create a derived value which is not just a slice of original atom's value, (i.e. `sum` field on above example ), do you think derived atom is needed or using select atom as perspective of using reducer? (since reducer creates a new data based on values/actions given)

**Reply by dai-shi** [reply, marked as answer]
I would discourage the use of `selectAtom`, and encourage derived atoms as much as possible.

```js
// recommended
const sumAtom = atom((get) => get(someAtom).a + get(someAtom).b)

// not recommended
const sumAtom = selectAtom(someAtom, (obj) => obj.a + obj.b)
```

**Reply by wootra** [reply]
Thanks for the quick answer like always!
